lib/utils/parse/ast.go

```diff
@@ -0,0 +1,487 @@
+/*
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package parse
+
+import (
+	"fmt"
+	"net/mail"
+	"reflect"
+	"regexp"
+	"strings"
+
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport"
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+// Expr is a node in the AST.
+type Expr interface {
+	// Kind indicates the expression kind.
+	Kind() reflect.Kind
+	// Evaluate evaluates the expression given the evaluation context.
+	Evaluate(ctx EvaluateContext) (any, error)
```

Comment:
v9 is still on go1.17, there is no `any`, need to explicitly write `interface{}`